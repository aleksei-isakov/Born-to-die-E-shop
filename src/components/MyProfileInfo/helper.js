export const MIN_NAME_LENGTH = 2;
export const MIN_PHONE_LENGTH = 10;

import { required, minLength, email, alpha } from 'vuelidate/lib/validators';

export const NAME_VALID = {
  required,
  alpha,
  minLength: minLength(MIN_NAME_LENGTH)
};

export const PHONE_VALID = {
  minLength: minLength(MIN_PHONE_LENGTH)
};

export const EMAIL_VALID = {
  required,
  email
};

export const formMixin = {
  methods: {
    onInputValidateField(fieldName) {
      const field = this.$v[fieldName];

      if (field) {
        return {
          'profile-form__error': field.$invalid && field.$dirty
        };
      }
    },

    onSubmitValidateForm() {
      this.$v.$touch();
    }
  }
};
