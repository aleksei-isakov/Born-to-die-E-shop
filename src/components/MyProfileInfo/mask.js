import { maskList } from './maskList';
const WHITE_SPACE = /[\s#]/g;
const WHITE_SPACES_AND_SYMBOLS = /[\s#-)(]/g;
const NOT_DIGIT = /\D/g;
const NON_PLUS_SYMBOLS = /(?!\+)./g;
const ANY_DIGIT = /[#\d]/;

export function setMask(inputValue) {
  let template = '+###############';

  maskList.forEach((mask) => {
    let code = mask.code.replace(WHITE_SPACE, '');
    let phone = inputValue.replace(WHITE_SPACES_AND_SYMBOLS, '');

    if (phone.includes(code)) {
      template = mask.code;
    }
  });

  let i = 0;
  let val = inputValue.replace(NOT_DIGIT, '');
  let formattedNumber = template.replace(NON_PLUS_SYMBOLS, function (a) {
    if (ANY_DIGIT.test(a) && i < val.length) {
      return val.charAt(i++);
    }

    if (i >= val.length) {
      return '';
    }

    return a;
  });

  return formattedNumber;
}
