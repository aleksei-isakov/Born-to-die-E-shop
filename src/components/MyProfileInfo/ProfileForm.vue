<template>
  <div class="profile-form">
    <div class="profile-form__category">{{ category }}</div>
    <input
      class="profile-form__input"
      :value="inputValue"
      @input="onInputChangeValue"
    />
    <span class="profile-form__error-message">{{ error }}</span>
  </div>
</template>

<script>
import { setMask } from './mask.js';
export default {
  name: 'ProfileForm',

  props: {
    category: {
      type: String,
      default: ''
    },

    error: {
      type: String,
      default: ''
    }
  },

  data() {
    return {
      inputValue: ''
    };
  },

  methods: {
    onInputChangeValue(event) {
      if (this.category === 'Phone number') {
        if (!this.inputValue) {
          this.inputValue = '+' + event.target.value;
        } else {
          this.inputValue = event.target.value;
        }
        this.inputValue = setMask(this.inputValue);
      } else {
        this.inputValue = event.target.value;
      }
      this.$emit('input', this.inputValue);
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';
.profile-form {
  position: relative;
  width: 80%;
  margin-top: 30px;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px;

  &__error {
    border: 1px solid $error-hover;

    .profile-form__error-message {
      color: $error-hover;
      position: absolute;
      display: block;
      margin-top: 3px;
      font-size: 12px;
    }
  }

  &__error-message {
    display: none;
  }

  &__input {
    width: 100%;
    border-radius: 5px;
    min-width: 120px;
    height: 5vh;
    padding: 0 10px;
  }
  &__category {
    background-color: $white;
    color: #b3b4b6;
    padding: 0 5px 0 5px;
    position: absolute;
    top: -8px;
    left: 30px;
    font-size: 12px;
  }
}
</style>
