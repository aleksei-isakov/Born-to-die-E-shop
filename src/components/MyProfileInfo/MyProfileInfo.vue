<template>
  <div v-if="currentUserInfo">
    <div>
      <form
        class="profile-info"
        novalidate
        @submit.prevent="onSubmitValidateForm"
      >
        <img class="profile-info__avatar" :src="defaultAvatar" />
        <profile-form
          category="First Name"
          :error="firstNameError"
          :class="onInputValidateField('firstName')"
          :user="currentUserInfo.user.firstName"
          @input="(inputValue) => onInputChangeValue('firstName', inputValue)"
        />
        <profile-form
          category="Last Name"
          :error="lastNameError"
          :class="onInputValidateField('lastName')"
          :user="currentUserInfo.user.lastName"
          @input="(inputValue) => onInputChangeValue('lastName', inputValue)"
        />
        <select-field
          ref="childComponent"
          category-name="Gender"
          :categories="genders"
          :index-cat="changeUserGenderCategory"
          class="profile-info__select"
          :default-category="1"
          @setGender="setGender"
        />
        <md-datepicker
          v-model="getDate"
          :md-open-on-focus="false"
          class="profile-info__calendar"
          :md-disabled-dates="isDateDisabled"
        >
          <div class="profile-info__date-category">Date of birth</div>
        </md-datepicker>
        <profile-form
          category="Phone number"
          :error="phoneError"
          :user="currentUserInfo.user.phoneNumber"
          :class="onInputValidateField('phoneNumber')"
          @input="(inputValue) => onInputChangeValue('phoneNumber', inputValue)"
        />
        <profile-form
          category="Email"
          :error="emailError"
          :user="currentUserInfo.user.email"
          :class="onInputValidateField('email')"
          @input="(inputValue) => onInputChangeValue('email', inputValue)"
        />
        <div class="profile-info__buttons">
          <base-button
            class="profile-info__button save"
            type="submit"
            @click="onClickChangeSaveNewUserInfo"
            >SAVE
          </base-button>
          <base-button class="profile-info__button cancel">CANCEL</base-button>
        </div>
      </form>
    </div>
  </div>
  <div v-else>
    <v-progress-circular indeterminate color="primary"></v-progress-circular>
  </div>
</template>

<script>
import defaultAvatar from '@/assets/defaultAvatar.png';
import ProfileForm from './ProfileForm.vue';
import SelectField from '@/components/SelectField/SelectField.vue';
import { BaseButton } from '@/base_components';
import { validationMixin } from 'vuelidate';
import { mapActions, mapGetters } from 'vuex';
import {
  EMAIL_VALID,
  formMixin,
  MIN_NAME_LENGTH,
  NAME_VALID,
  PHONE_VALID
} from './helper';

let GENDERS = ['not specified', 'male', 'female'];

export default {
  name: 'MyProfileInfo',

  components: {
    ProfileForm,
    BaseButton,
    SelectField
  },

  mixins: [validationMixin, formMixin],

  data() {
    return {
      defaultAvatar: defaultAvatar,
      email: '',
      firstName: '',
      lastName: '',
      phoneNumber: '',
      userGender: '',
      genders: GENDERS,

      validData: {
        hasFirstName: false,
        hasLastName: false,
        hasEmail: false,
        hasPhoneNumber: false
      },

      selectedDate: null
    };
  },

  computed: {
    ...mapGetters('AuthenticationModule', ['currentUserInfo']),

    getDate: {
      set: function (val) {
        this.selectedDate = new Date(val);
      },

      get: function () {
        return new Date(this.currentUserInfo.user.birthDate);
      }
    },

    firstNameError() {
      if (!this.$v.firstName.required) {
        this.failedValidData('hasFirstName');

        return 'The first name is required';
      }

      if (!this.$v.firstName.minLength || !this.$v.firstName.alpha) {
        this.failedValidData('hasFirstName');

        return `Must contain only latin letters, at least ${MIN_NAME_LENGTH} symbols`;
      }

      return this.successValidData('hasFirstName');
    },

    lastNameError() {
      if (!this.$v.lastName.required) {
        this.failedValidData('hasLastName');

        return 'The last name is required';
      }

      if (!this.$v.lastName.minLength || !this.$v.lastName.alpha) {
        this.failedValidData('hasLastName');

        return `Must contain only latin letters, at least ${MIN_NAME_LENGTH} symbols`;
      }

      return this.successValidData('hasLastName');
    },

    emailError() {
      if (!this.$v.email.required) {
        this.failedValidData('hasEmail');

        return 'The email is required';
      }

      if (!this.$v.email.email) {
        this.failedValidData('hasEmail');

        return 'Invalid email';
      }

      return this.successValidData('hasEmail');
    },

    phoneError() {
      if (!this.$v.phoneNumber.minLength) {
        this.failedValidData('hasPhoneNumber');

        return 'Too short number';
      }

      return this.successValidData('hasPhoneNumber');
    },

    changeUserGenderCategory() {
      return this.genders.indexOf(this.currentUserInfo.user.gender);
    }
  },

  methods: {
    ...mapActions('AuthenticationModule', ['getUserInfo', 'updateUserInfo']),

    onInputChangeValue(category, inputValue) {
      this[category] = inputValue;
    },

    onClickChangeSaveNewUserInfo() {
      this.email = this.email || this.currentUserInfo.user.email;
      this.firstName = this.firstName || this.currentUserInfo.user.firstName;
      this.lastName = this.lastName || this.currentUserInfo.user.lastName;
      this.phoneNumber =
        this.phoneNumber || this.currentUserInfo.user.phoneNumber;

      if (this.phoneNumber === '+') {
        this.phoneNumber = '';
      }

      this.updateUserInfo({
        email: this.email || this.currentUserInfo.user.email,
        firstName: this.firstName || this.currentUserInfo.user.firstName,
        lastName: this.lastName || this.currentUserInfo.user.lastName,
        phoneNumber: this.phoneNumber,
        gender: this.userGender || this.currentUserInfo.user.gender,
        birthDate:
          new Date(this.selectedDate) || this.currentUserInfo.user.birthDate
      });
    },

    failedValidData(hasCategory) {
      this.validData[hasCategory] = false;
    },

    successValidData(hasCategory) {
      this.validData[hasCategory] = true;
    },

    isDateDisabled(date) {
      return new Date() < date;
    },

    setGender(gender) {
      this.userGender = gender;
    }
  },

  validations: {
    firstName: NAME_VALID,
    lastName: NAME_VALID,
    email: EMAIL_VALID,
    phoneNumber: PHONE_VALID
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.profile-info {
  max-width: 700px;
  padding: 25px 80px 10px 80px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  margin: 0 auto;

  @media screen and (max-width: $tablet-size) {
    padding: 20px;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
  }

  .sign-in__error {
    text-align: center;
    color: $error;
  }

  &__calendar {
    width: 100%;
    border: 1px solid rgb(211, 210, 210);
    border-radius: 5px;
    margin: 30px 0 0 0;
    padding: 0;
    height: 6vh;
    align-items: center;
    position: relative;
    z-index: 1;
    max-width: 450px;
    max-height: 50px;
    min-height: 0;

    &.md-field::v-deep .md-button {
      top: 2px;
    }

    &.md-field::after::v-deep {
      height: 0;
    }

    &.md-field::v-deep .md-icon::after {
      height: 0;
    }

    .profile-info__date-category {
      position: absolute;
      background-color: $white;
      color: #b3b4b6;
      padding: 0 5px;
      font-size: 12px;
      top: -8px;
      left: 30px;
    }
  }

  &__select {
    width: 100%;
    position: relative;
    z-index: 5;
    margin-top: 30px;
    max-width: 450px;

    &::v-deep .select-field__category {
      left: 30px;
    }

    &::v-deep .select-field {
      width: 100%;
      padding: 0 0 0 inherit;
      margin: 30px 0 0 0;
    }

    &::v-deep .select-field__button {
      height: 6vh;
      max-height: 50px;
    }

    &::v-deep .dropdown {
      width: 100%;
      position: absolute;
    }
  }

  &__avatar {
    border-radius: 50%;
    width: 20vh;
    height: 20vh;
    min-width: 100px;
    min-height: 100px;
  }

  .profile-info__buttons {
    margin: 10% 0;

    .profile-info__button {
      width: 100px;
      border: 0.5px solid $font-color-text;
      padding: 5px 0px;
      margin: 0px 10px;
      border-radius: 5px;
    }

    .save {
      color: $primary;
    }

    .cancel {
      color: $error;
    }
  }
}
</style>
