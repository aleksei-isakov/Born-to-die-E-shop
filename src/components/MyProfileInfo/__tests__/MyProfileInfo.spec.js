import MyProfileInfo from '../MyProfileInfo.vue';
import { shallowMount } from '@vue/test-utils';

describe('MyProfileInfo', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(MyProfileInfo, {
      stubs: { 'md-datepicker': true }
    });

    expect(wrapper.element).toMatchSnapshot();
  });
  it('shows first name error', () => {
    const wrapper = shallowMount(MyProfileInfo, {
      data() {
        return { firstName: '' };
      },
      stubs: { 'md-datepicker': true }
    });

    expect(wrapper.vm.firstNameError).toBe('The first name is required');
  });

  it('shows last name error', () => {
    const wrapper = shallowMount(MyProfileInfo, {
      data() {
        return { lastName: '1' };
      },
      stubs: { 'md-datepicker': true }
    });

    expect(wrapper.vm.lastNameError).toBe(
      'Must contain only latin letters, at least 2 symbols'
    );
  });

  it('shows email error', () => {
    const wrapper = shallowMount(MyProfileInfo, {
      data() {
        return { email: 'email' };
      },
      stubs: { 'md-datepicker': true }
    });

    expect(wrapper.vm.emailError).toBe('Invalid email');
  });
});
