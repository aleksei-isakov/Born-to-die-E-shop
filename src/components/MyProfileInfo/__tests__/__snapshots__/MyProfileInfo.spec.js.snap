// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyProfileInfo renders correctly 1`] = `
<div>
  <div>
    <form
      class="profile-info"
      novalidate="novalidate"
    >
      <img
        class="profile-info__avatar"
        src=""
      />
       
      <profile-form-stub
        category="First Name"
        class=""
        error="The first name is required"
        user="Admin"
      />
       
      <profile-form-stub
        category="Last Name"
        class=""
        error="The last name is required"
        user="Adminson"
      />
       
      <select-field-stub
        categories="not specified,male,female"
        category-name="Gender"
        class="profile-info__select"
        default-category="1"
        indexcat="1"
      />
       
      <md-datepicker-stub
        class="profile-info__calendar"
        md-disabled-dates="function () { [native code] }"
        value="Invalid Date"
      >
        <div
          class="profile-info__date-category"
        >
          Date of birth
        </div>
      </md-datepicker-stub>
       
      <profile-form-stub
        category="Phone number"
        class=""
        error=""
        user="+79119229222"
      />
       
      <profile-form-stub
        category="Email"
        class=""
        error="The email is required"
        user="admin@test.com"
      />
       
      <div
        class="profile-info__buttons"
      >
        <base-button-stub
          class="profile-info__button save"
          type="submit"
        >
          SAVE
        
        </base-button-stub>
         
        <base-button-stub
          class="profile-info__button cancel"
        >
          CANCEL
        </base-button-stub>
      </div>
    </form>
  </div>
</div>
`;
